File: leetcode/3Sum.js
/**
 * @param {interger} nums 
 */
var twoSum = function (target, nums) {
    result = []
    numbsObj = {}

    /**
     * O(N) time complexity and the space complexity is O(N)
     * */
    for (let num of nums) {
        // calculating the remainder
        let compliment = target - num;
        if (compliment in numbsObj) {
            result.push([num, compliment])
        }
        numbsObj[num] = 0
    }
    return result
}

/**
 * @param {integer} nums 
 * @returns list of triplets that sums upto 0
 * @description  The approach used here is to use a number and then access the next two ones byb adding the latter and finally adding
 * with former to get 0.
 */
var threeSum = function (nums) {
    /**
     * refer Dynamic type for structural typing
     */
    if (nums.constructor !== Array) {
        return "Illegal type passed"
    }
    output = []
    nums.sort((a, b) => a - b)
    for (var i = 0; i < nums.length - 2; i++) {
        if (i === 0 || (i > 0 && nums[i] != nums[i - 1])) {
            // 2 pointers are incremented/decremented to perform combinations on given ranges
            let low = i + 1
            let high = nums.length - 1
            // here 0 is the target and the complement defines the target for addition of 2 numbers  
            let sum = 0 - nums[i]
            // moving inward starting from the 2 pointers
            while (low < high) {
                // checking if the 2 points match the sum 
                if (nums[low] + nums[high] === sum) {
                    // match found 
                    output.push([nums[i], nums[low], nums[high]])
                    /**
                     * avoiding duplicates by changing the position of pointers 
                     * low < high : within a given range
                     * nums[low] === nums[low+1] : values of the starting pointer and the next pointer are one and the same 
                     *                             then increment pointer to next position
                     * nums[high] === nums[high -1]: values of ending pointer and the previous pointer are one and the same
                     *                               then decrement pointer to previous position
                     * 
                     * */
                    while (low < high && nums[low] === nums[low + 1])
                        low++;
                    while (low < high && nums[high] === nums[high - 1])
                        high--;
                    // both values at the pointers failed to achieve target sum thereby moving the pointers inwards
                    low++;
                    high--;
                }
                // due to sorting the right hand side will hold all the positive numbers, left hand side will hold all the negative numbers given that array is a sequence of integers
                else if (nums[low] + nums[high] > sum)
                    high--;
                else
                    low++;
            }
        }
    }
    return output
}

// console.log(twoSum(10, [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]))
// console.log(threeSum([-4, -1, -1, 0, 1, 2]))
console.log(threeSum([-4, -2, -2, -1, -1, 0, 1, 2, 3, 4]))

File: leetcode/3sum.py
def three_sum(arr, target=0):
    output = []
    # sorting the arr so the search becomes easy
    sorted_arr = sorted(arr)
    for ele in range(len(sorted_arr) - 2):
        if (ele == 0 or (ele > 0 and sorted_arr[ele] != sorted_arr[ele - 1])):

            low = ele + 1
            high = len(sorted_arr) - 1

            sum_to_get = target - sorted_arr[ele]
            while(low < high):
                if (sorted_arr[low] + sorted_arr[high] == sum_to_get):
                    output.append(
                        [
                            sorted_arr[ele],
                            sorted_arr[low],
                            sorted_arr[high]
                        ]
                    )
                    while(low < high and sorted_arr[low] == sorted_arr[low + 1]):
                        low += 1
                    while(low < high and sorted_arr[high] == sorted_arr[high - 1]):
                        high -= 1
                    low += 1
                    high -= 1
                elif (sorted_arr[low] + sorted_arr[high] > sum_to_get):
                    high -= 1
                else:
                    low += 1
    return output


# print(three_sum([-1, 0, 1, 2, -1, -4])) # [[-1, -1, 2], [-1, 0, 1]]

# [[-4, 0, 4], [-4, 1, 3], [-2, -2, 4], [-2, -1, 3], [-2, 0, 2], [-1, -1, 2], [-1, 0, 1]]
# print(three_sum([-4, -2, -2, -1, -1, 0, 1, 2, 3, 4]))


def three_sum_closest(arr, target=1):
    comb_sum_set = set()
    sorted_arr = sorted(arr)
    closest_value = None

    """
    .Logic:
        ..fetching all the combinations that sums up to a number
        ..After getting the list of sums from the combinations find the `ranges` of these sums and then 
          selecting the corresponding closest number out of these
    """
    #  Combinational Sums calculation
    for ele in range(len(sorted_arr) - 2):
        # based on the number of integers
        low = ele + 1
        high = len(sorted_arr) - 1
        while(low < high):
            total = sorted_arr[ele] + sorted_arr[low] + sorted_arr[high]
            comb_sum_set.add(total)

            while(low < high and sorted_arr[low] == sorted_arr[low + 1]):
                low += 1
            while(low < high and sorted_arr[high] == sorted_arr[high - 1]):
                high -= 1

            low += 1
            # high -= 1

    # after getting the Combinal Sum Set, find the intervals between the numbers in set and target
    import math
    min = math.inf
    sum_scale = {}

    for i in comb_sum_set:
        if i == target:
            return i
        count = len(range(i, target))
        sum_scale[i] = count

    for key, value in sum_scale.items():
        if value < min:
            min = value
            closest_value = key
    return closest_value


# print(three_sum_closest([-1, 0, 1, 2, -1, -4]))
# print(three_sum_closest([0, 2, 1, -3]))


def sum_of_three_closest(arr, target=1):
    sorted_arr = sorted(arr)
    sorted_arr_len = len(sorted_arr)
    result = sorted_arr[0] + sorted_arr[1] + sorted_arr[sorted_arr_len - 1]
    for i in range(sorted_arr_len):
        j, k = i+1, sorted_arr_len - 1
        while j < k:
            current = sorted_arr[i] + sorted_arr[j] + sorted_arr[k]
            if current == target:
                return target
            elif abs(current - target) < abs(result - target):
                result = current
            elif current > target:
                k -= 1
            else:
                j += 1
    return result


print(sum_of_three_closest([0, 2, 1, -3]))
